### ISSUE TITLE
Question regarding GRPOTrainer's advantage clipping

### ISSUE THREAD

**Issue by haikangdeng** [huggingface/trl#3757]
### Reproduction

The default `num_iterations` in GRPOTrainer is set to 1. As a result, `coef_1` and `coef_2` are always`1` in trl/trainer/grpo_trainer.py, which means no clipping will occur. Is this expected behavior?

For every new batch, when the num_iterations is set to 1, should $\pi_{old}$ be the last policy model or current policy model? Current implementation seems to set $\pi_{old}$ to the current model.

trl/trainer/grpo_trainer.py [Line1506-1510](https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L1506-L1510)
```python
    old_per_token_logps = (
        per_token_logps.detach() if inputs["old_per_token_logps"] is None else inputs["old_per_token_logps"]
    )
    coef_1 = torch.exp(per_token_logps - old_per_token_logps)
    coef_2 = torch.clamp(coef_1, 1 - self.epsilon_low, 1 + self.epsilon_high)
```

Wandb figures indicate that no clipping ever happened throughout the training.

<img width="736" height="581" alt="Image" src="https://github.com/user-attachments/assets/fcc378d2-d484-4c67-9ebb-7a97e733fa3e" />

### System Info

``` 
- Platform: Linux-6.8.0-41-generic-x86_64-with-glibc2.39
- Python version: 3.10.14
- TRL version: 0.20.0.dev0+e333da8
- PyTorch version: 2.7.0
- accelerator(s): NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3
- Transformers version: 4.53.2
- Accelerate version: 1.7.0
- Accelerate config:
  - compute_environment: LOCAL_MACHINE
  - distributed_type: MULTI_GPU
  - mixed_precision: fp16
  - use_cpu: False
  - debug: False
  - num_processes: 3
  - machine_rank: 0
  - num_machines: 1
  - gpu_ids: all
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 3.2.0
- HF Hub version: 0.33.4
- bitsandbytes version: 0.46.0
- DeepSpeed version: 0.17.2
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.88.0
- PEFT version: not installed
- vLLM version: 0.9.2
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by ankur6ue** [comment]
TRL's API for on-policy/off-policy training using PPO/GRPO etc., is unnecessarily confusing I think. They have two parameters steps_per_generation and num_iterations, when only one is needed. See my related post:
https://github.com/huggingface/trl/issues/3662

It has been 3 weeks, and no answers yet..

**Comment by pramodith** [comment]
@haikangdeng if `num_iterations`=1, the policy loss will always be zero since the current and old policy are the same model. I wrote about this in depth in my [blog](https://pramodith.github.io/posts/grpo-trainer/#curious-case-of-number-of-iterations--1).

You're right about the clipping factors also ending up to be 1 as a result. The thing is that even though the loss is 0 the gradients aren't so the model still learns. For whatever, reason setting `num_iterations` to 1 has become the de-facto in GRPO and it seems to work. I have my own theories around why this is a sub-optimal setting but I'd encourage you to play around with the hyper-param to see if you can get better results.

**Comment by avishaiElmakies** [comment]
@ankur6ue I have tried to improve the Docs and created a PR, would love to hear what you think.
`steps_per_generation` and `num_iterations` are not the same thing, and this is why i created the PR. I agree the current Docs are confusing.
https://github.com/huggingface/trl/pull/3761

**Comment by qgallouedec** [comment]
Hey @haikangdeng, sorry for the long delay,catching up on the backlog.

This is expected behavior, not a bug. With `num_iterations=1` (and gradient-accumulation steps aligned with the generation cycle), each optimizer step uses freshly generated completions whose log-probs were computed with the same weights that are about to be updated, so by construction `old_per_token_logps == per_token_logps.detach()` and the ratio is exactly 1 (no clipping fires). `π_old` only diverges from `π` once you do >1 update per generated batch, i.e. `num_iterations > 1`, or when the steps are misaligned / vLLM importance sampling is on, in which case we explicitly recompute `old_per_token_logps` (see https://github.com/huggingface/trl/blob/a0d909b1a099b4cf52ee12648142a4cd7d39ca04/trl/trainer/grpo_trainer.py#L2038-L2061). This is the same question as #3294, where @LeonEricsson walked through the iteration semantics in detail — definitely worth a read. Also +1 to @pramodith's blog and @avishaiElmakies's docs PR #3761.
